File: dart/pascals-triangle/lib/pascals_triangle.dart
class PascalsTriangle {
  List<List<int>> rows(int row) {
    return List.generate(row, (index) => index).fold([],
        (List<List<int>> previousValue, int element) {
      return [...previousValue, _calcNextRow(previousValue, element)];
    });
  }

  List<int> _calcNextRow(List<List<int>> triangle, int row) {
    if (row == 0) {
      return [1];
    }

    return List.generate(row + 1, (index) {
      if (index == 0 || index == row) {
        return 1;
      }

      return triangle[row - 1][index - 1] + triangle[row - 1][index];
    });
  }
}

File: kotlin/pascals-triangle/src/main/kotlin/PascalsTriangle.kt
object PascalsTriangle {

    fun computeTriangle(rows: Int): List<List<Int>> {
        val triangle = Array(rows) { IntArray(rows) }

        for (i in triangle.indices) {
            for (j in 0..i) {
                when {
                    j == 0 -> triangle[i][j] = 1
                    j == i -> triangle[i][j] = 1
                    else -> triangle[i][j] = triangle[i - 1][j] + triangle[i - 1][j - 1]
                }
            }
        }

        return triangle.map { it.toList().filter { it -> it != 0 } }
    }
}
